pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java

```diff
@@ -362,7 +363,12 @@ private void readMoreEntries(Consumer consumer) {
                 }
                 havePendingRead = true;
                 if (consumer.readCompacted()) {
-                    boolean readFromEarliest = isFirstRead && MessageId.earliest.equals(consumer.getStartMessageId());
+                    boolean readFromEarliest = false;
+                    if (!cursor.isDurable() || ((ManagedCursorImpl) cursor).isCompactionCursor()
+                            || cursor.getMarkDeletedPosition() == null
+                            || cursor.getMarkDeletedPosition().getEntryId() == -1L) {
+                        readFromEarliest = isFirstRead && MessageId.earliest.equals(consumer.getStartMessageId());
+                    }
```

Comment:
Could you just simplify the check with:

```java
                    boolean readFromEarliest = isFirstRead && cursor.isDurable()
                            && cursor.getName().equals(Compactor.COMPACTION_SUBSCRIPTION);
```

You don't need to expose the `ManagedCursor#isCompactionCursor()` method.

And why did you also check the mark deleted position here?

Reply:
> And why did you also check the mark deleted position here?

If the mark-delete position is null and `subscriptionInitialPosition=Earliest` then we still need to read messages from earliest.

Reply:
> If the mark-delete position is null and subscriptionInitialPosition=Earliest

Could you add a test for it?

Reply:
`testAcknowledgeWithReconnection` already covers this case.

Reply:
I removed the check for mark-delete position and `testAcknowledgeWithReconnection` still passed. So I think it's not covered.

```java
                    boolean readFromEarliest = isFirstRead && MessageId.earliest.equals(consumer.getStartMessageId())
                            && (!cursor.isDurable() || cursor.getName().equals(Compactor.COMPACTION_SUBSCRIPTION));
```

Reply:
Yeah, after unloading the topic, also need to write a new message otherwise `ml.getFirstPosition()` returns the old ledger position, I added `testEarliestSubsAfterRollover` to cover this case.